File: reader/src/sync/merge.ts
import type { Annotations, Positions } from "@scriptorium/shared";

// A bit-for-bit TypeScript port of `server/src/scriptorium/sync/merge.py` (DESIGN §12). The server is
// authoritative — R3's engine adopts the server's merged doc wholesale — but the sync engine also
// canonicalizes the LOCAL doc before PUT (dedup / tombstone hygiene while offline), so this port must
// converge to exactly the same result the server would. The two impls are pinned against ONE vector
// file (`shared/test-vectors/sync-merge.json`), consumed by `merge.test.ts` here and
// `server/tests/test_sync_vectors.py`; a drift reddens one of the two suites.
//
// Timestamps are compared as ISO-8601 STRINGS, never parsed to Date: they are always UTC and
// second-or-finer precision (JS `Date.toISOString()`), so lexicographic order equals chronological
// order. Do not "improve" this to Date parsing — string compare is the contract.

type Annotation = Annotations["annotations"][number];
type Position = Positions["current"];

/**
 * Reproduce Python's `json.dumps(value, sort_keys=True, ensure_ascii=False)` for the annotation value
 * domain (objects, strings, integers, booleans). This is the deterministic tiebreak used when two
 * copies of an annotation share an `id` AND a `modified` timestamp, so it must match the server byte
 * for byte: object keys sorted, `", "` / `": "` separators, non-ASCII left raw. A naive
 * `JSON.stringify` (no spaces, escapes non-ASCII) would diverge on that degenerate case.
 */
export function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") {
    // Scalars: JSON.stringify already matches Python for strings (raw non-ASCII), ints, and booleans.
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(canonicalJson).join(", ")}]`;
  }
  const keys = Object.keys(value as Record<string, unknown>).sort();
  const body = keys
    .map((k) => `${JSON.stringify(k)}: ${canonicalJson((value as Record<string, unknown>)[k])}`)
    .join(", ");
  return `{${body}}`;
}

/** Which of two same-`id` annotations wins: greater `modified`, then full-doc canonical-JSON tiebreak. */
function annotationPickKey(ann: Annotation): [string, string] {
  return [ann.modified, canonicalJson(ann)];
}

/** Total order on the pick key: greater `modified`, then greater canonical JSON (both string compares). */
function pickKeyGreater(a: [string, string], b: [string, string]): boolean {
  if (a[0] !== b[0]) return a[0] > b[0];
  return a[1] > b[1];
}

/** Dedup a doc's annotations by `id` (winner kept) and sort ascending by `id` — the canonical form. */
function canonicalAnnotations(doc: Annotations): Annotations {
  const best = new Map<string, Annotation>();
  for (const ann of doc.annotations) {
    const cur = best.get(ann.id);
    if (cur === undefined || pickKeyGreater(annotationPickKey(ann), annotationPickKey(cur))) {
      best.set(ann.id, ann);
    }
  }
  const ordered = [...best.values()].sort((x, y) => (x.id < y.id ? -1 : x.id > y.id ? 1 : 0));
  return { book_id: doc.book_id, user_id: doc.user_id, annotations: ordered };
}

/**
 * Merge two annotation docs (DESIGN §12): union by `id`, last-writer-wins by `modified`. Tombstones
 * (`deleted: true`) are ordinary entries — a later delete beats an earlier edit and vice versa
 * (deletion can lose; documented). Output is canonical (deduped, sorted by `id`). Identity comes from
 * `a` (falling back to `b`). Commutative, associative, idempotent (`merge(x, x) == canonical(x)`).
 */
export function mergeAnnotations(a: Annotations, b: Annotations): Annotations {
  const identity = a.book_id ? a : b;
  const combined: Annotations = {
    book_id: identity.book_id,
    user_id: identity.user_id,
    annotations: [...a.annotations, ...b.annotations],
  };
  return canonicalAnnotations(combined);
}

/** `furthest` order: by `(page_seq, char)` — furthest-read-wins; `modified` only tiebreaks a tie. */
function furthestGreater(a: Positions["furthest"], b: Positions["furthest"]): boolean {
  if (a.page_seq !== b.page_seq) return a.page_seq > b.page_seq;
  if (a.char !== b.char) return a.char > b.char;
  return a.modified > b.modified;
}

/** `current` order: LWW by `modified`, then `(page_seq, char, device)` as a deterministic tiebreak. */
function currentGreater(a: Position, b: Position): boolean {
  if (a.modified !== b.modified) return a.modified > b.modified;
  if (a.page_seq !== b.page_seq) return a.page_seq > b.page_seq;
  if (a.char !== b.char) return a.char > b.char;
  return (a.device ?? "") > (b.device ?? "");
}

/**
 * Merge two position docs (DESIGN §12): `furthest` = max by `(page_seq, char)` regardless of
 * timestamp (never regresses); `current` = last-writer-wins by `modified`. Commutative, associative,
 * idempotent.
 */
export function mergePositions(a: Positions, b: Positions): Positions {
  return {
    furthest: furthestGreater(b.furthest, a.furthest) ? b.furthest : a.furthest,
    current: currentGreater(b.current, a.current) ? b.current : a.current,
  };
}

File: server/src/scriptorium/sync/merge.py
"""Server-authoritative sync merge — the convergence core (DESIGN §12).

Pure and FastAPI-free: clients send full documents and receive merged full documents, and these two
functions decide the outcome. Correctness rests on **order-independence** — two offline devices may
reconnect in either order, possibly repeatedly, and must converge to the same document. That is why
both functions are commutative, associative, and idempotent, proven by the property harness in
``tests/test_sync_merge.py``.

The functions never do I/O and never call the network; the endpoint layer (``sync/api.py``) wraps
them with schema validation, per-``(user, book)`` locking, and versioned backups.

Timestamps are compared as **ISO-8601 strings**: they are always UTC and always second-or-finer
precision (JS ``Date.toISOString()``), so lexicographic order equals chronological order — no
parsing needed, and a malformed-but-comparable string still yields a deterministic result.
"""

from __future__ import annotations

import json
from typing import Any

# Tombstones (deleted annotations) are retained at least this long so a device that was offline for
# a while still sees the deletion and converges. Actually *dropping* aged tombstones (compaction) is
# explicitly deferred: v1 keeps every tombstone forever, so this constant is documentation of the
# floor a future compactor must honor, not a value the merge currently enforces.
TOMBSTONE_RETENTION_DAYS = 180


def _annotation_sort_key(ann: dict[str, Any]) -> str:
    """Total, stable ordering of annotations by their merge key (``id``)."""
    return ann["id"]


def _annotation_pick_key(ann: dict[str, Any]) -> tuple[str, str]:
    """Which of two same-``id`` annotations wins: greater ``modified``, then a full-doc tiebreak.

    The tiebreak (canonical JSON of the whole annotation) only matters when two copies share an id
    *and* a ``modified`` timestamp; it guarantees the choice is independent of argument order, so
    the merge stays commutative even in that degenerate case.
    """
    return (ann["modified"], json.dumps(ann, sort_keys=True, ensure_ascii=False))


def _canonical_annotations(doc: dict[str, Any]) -> dict[str, Any]:
    """Return ``doc`` with its annotations deduplicated by id (winner kept) and sorted by id.

    Idempotency of :func:`merge_annotations` is defined against this canonical form — merging a
    document with itself yields exactly ``_canonical_annotations`` of it.
    """
    best: dict[str, dict[str, Any]] = {}
    for ann in doc.get("annotations", []):
        cur = best.get(ann["id"])
        if cur is None or _annotation_pick_key(ann) > _annotation_pick_key(cur):
            best[ann["id"]] = ann
    ordered = sorted(best.values(), key=_annotation_sort_key)
    return {
        "book_id": doc["book_id"],
        "user_id": doc["user_id"],
        "annotations": ordered,
    }


def merge_annotations(a: dict[str, Any], b: dict[str, Any]) -> dict[str, Any]:
    """Merge two annotation docs (DESIGN §12): union by ``id``, last-writer-wins by ``modified``.

    Tombstones (``deleted: true``) are ordinary entries — a deletion is just another edit and merges
    identically, so a later delete beats an earlier edit and a later edit beats an earlier delete
    (deletion can lose; documented and acceptable). Output is canonical: annotations sorted by
    ``id``, each id present once. Identity comes from ``a`` (falling back to ``b``); callers that
    know the ``(user, book)`` from the request path overwrite both fields authoritatively afterward.

    Commutative, associative, and idempotent (``merge(x, x) == _canonical_annotations(x)``).
    """
    identity = a if a.get("book_id") else b
    combined = {
        "book_id": identity["book_id"],
        "user_id": identity["user_id"],
        "annotations": list(a.get("annotations", [])) + list(b.get("annotations", [])),
    }
    return _canonical_annotations(combined)


def _furthest_key(pos: dict[str, Any]) -> tuple[int, int, str]:
    """Order for ``furthest``: by ``(page_seq, char)`` — furthest-read-wins. ``modified`` is only a
    deterministic tiebreak when the reading point is identical, so it can never flip the winner."""
    return (pos["page_seq"], pos["char"], pos["modified"])


def _current_key(pos: dict[str, Any]) -> tuple[str, int, int, str]:
    """Order for ``current``: last-writer-wins by ``modified``, with ``(page_seq, char, device)`` as
    a deterministic tiebreak so equal-timestamp writes (even differing only by ``device``) still
    merge order-independently."""
    return (pos["modified"], pos["page_seq"], pos["char"], pos.get("device", ""))


def merge_positions(a: dict[str, Any], b: dict[str, Any]) -> dict[str, Any]:
    """Merge two position documents (DESIGN §12).

    ``furthest`` = max by ``(page_seq, char)`` **regardless of timestamp** (furthest-read-wins, so
    it never regresses); ``current`` = greater ``modified`` (last-writer-wins). Commutative,
    associative, and idempotent.
    """
    furthest = max(a["furthest"], b["furthest"], key=_furthest_key)
    current = max(a["current"], b["current"], key=_current_key)
    return {"furthest": furthest, "current": current}
